src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/ChainPal.cs

```diff
@@ -35,7 +38,319 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
+                    if (customTrustCerts.Count == 0)
+                    {
+                        throw new PlatformNotSupportedException(SR.Chain_EmptyCustomTrustNotSupported);
+                    }
+
+                    IntPtr[] customTrustArray = customTrustCerts.ToArray();
+                    int res = Interop.AndroidCrypto.X509ChainSetCustomTrustStore(_chainContext, customTrustArray, customTrustArray.Length);
+                    if (res != 1)
+                    {
+                        throw new CryptographicException();
+                    }
+                }
+            }
+
+            internal void Evaluate(
+                DateTime verificationTime,
+                OidCollection applicationPolicy,
+                OidCollection certificatePolicy,
+                X509RevocationMode revocationMode,
+                X509RevocationFlag revocationFlag)
+            {
+                Debug.Assert(_chainContext != null);
+
+                long timeInMsFromUnixEpoch = new DateTimeOffset(verificationTime).ToUnixTimeMilliseconds();
+                _isValid = Interop.AndroidCrypto.X509ChainBuild(_chainContext, timeInMsFromUnixEpoch);
+                if (!_isValid)
+                {
+                    // Android always validates name, time, signature, and trusted root.
+                    // There is no way bypass that validation and build a path.
+                    ChainElements = Array.Empty<X509ChainElement>();
+
+                    Interop.AndroidCrypto.ValidationError[] errors = Interop.AndroidCrypto.X509ChainGetErrors(_chainContext);
+                    var chainStatus = new X509ChainStatus[errors.Length];
+                    for (int i = 0; i < errors.Length; i++)
+                    {
+                        Interop.AndroidCrypto.ValidationError error = errors[i];
+                        chainStatus[i] = ValidationErrorToChainStatus(error);
+                        Marshal.FreeHGlobal(error.Message);
+                    }
+
+                    ChainStatus = chainStatus;
+                    return;
+                }
+
+                if (revocationMode != X509RevocationMode.NoCheck)
+                {
+                    if (revocationFlag == X509RevocationFlag.EntireChain)
+                    {
+                        throw new NotImplementedException($"{nameof(Evaluate)} (X509RevocationFlag.{revocationFlag})");
+                    }
+
+                    if (!Interop.AndroidCrypto.X509ChainSupportsRevocationOptions())
+                    {
+                        if (revocationFlag == X509RevocationFlag.EndCertificateOnly)
+                        {
+                            // No way to specfiy end certificate only if revocation options are not available
+                            throw new PlatformNotSupportedException(SR.Format(SR.Chain_SettingNotSupported, $"{nameof(X509RevocationFlag)}.{nameof(X509RevocationFlag.EndCertificateOnly)}"));
+                        }
```

Comment:
Do we get to learn what level was revoked?  If so, the fallback would be "build the chain with revocation on, if revocation occurred past the end-entity certificate, build it again with revocation off, if there are no errors, everything's cool".  (Or, if revocation is guaranteed to be the last thing reported we could just ignore it at levels beyond the end-entity cert)

Reply:
We do get the index, but the validation throws at the first error and the checker does its checks from most-trusted to target, so we don't know that the end cert actually passes the revocation check if an intermediate failed the revocation check.

Reply:
Oh, yep.  Right.  Build from bottom up, check revocation from top down (because if the middle is revoked then you can't actually trust the CRL that says the next level isn't revoked).

While there are some tests that care, generally specifying EndCertificateOnly is just about a perf win.  On the versions that don't support it I think we can just let all three levels be evaluated as ExcludeRoot.  (Basically, consider it as "we set the option, but the underlying implementation ignored it"... we did the best we could.  Not a huge problem because we're failing in the safer direction.)

Reply:
Treating `EndCertificateOnly` as `ExcludeRoot` now when we don't have the options.